File: tools/lib/error-handler.js
/**
 * MCP Error Handler
 * 
 * Standardized error handling for MCP client tools.
 * Provides consistent error types, formatting, and recovery mechanisms.
 */

/**
 * Standard error types for MCP client tools
 * @enum {string}
 */
export const ErrorType = {
  // Connection Errors
  CONNECTION_FAILED: 'connection_failed',
  TIMEOUT: 'timeout',
  SERVER_UNAVAILABLE: 'server_unavailable',
  
  // Tool Errors
  TOOL_NOT_FOUND: 'tool_not_found',
  TOOL_EXECUTION_FAILED: 'tool_execution_failed',
  
  // Parameter Errors
  INVALID_PARAMETERS: 'invalid_parameters',
  MISSING_REQUIRED_PARAMETER: 'missing_required_parameter',
  
  // Data Errors
  INVALID_RESPONSE: 'invalid_response',
  PARSING_ERROR: 'parsing_error',
  
  // File System Errors
  FILE_NOT_FOUND: 'file_not_found',
  PERMISSION_DENIED: 'permission_denied',
  
  // Session Errors
  SESSION_EXPIRED: 'session_expired',
  SESSION_CREATION_FAILED: 'session_creation_failed',
  
  // Generic Errors
  UNKNOWN: 'unknown_error',
  INTERNAL: 'internal_error'
};

/**
 * Severity levels for errors
 * @enum {string}
 */
export const ErrorSeverity = {
  FATAL: 'fatal',     // Application cannot continue
  ERROR: 'error',     // Operation failed but application can continue
  WARNING: 'warning', // Operation succeeded with issues
  INFO: 'info'        // Informational message
};

/**
 * Standard MCP Error class
 */
export class MCPError extends Error {
  /**
   * Create a new MCP Error
   * @param {string} message - Human-readable error message
   * @param {string} type - Error type from ErrorType enum
   * @param {Object} options - Additional error options
   * @param {string} [options.severity=ErrorSeverity.ERROR] - Error severity
   * @param {boolean} [options.retryable=false] - Whether the operation can be retried
   * @param {number} [options.code=500] - Error code
   * @param {Object} [options.details={}] - Additional error details
   * @param {string} [options.suggestion] - Suggested solution for the error
   */
  constructor(message, type, options = {}) {
    super(message);
    this.name = 'MCPError';
    this.type = type || ErrorType.UNKNOWN;
    this.severity = options.severity || ErrorSeverity.ERROR;
    this.retryable = options.retryable || false;
    this.code = options.code || 500;
    this.details = options.details || {};
    this.suggestion = options.suggestion;
    this.timestamp = new Date().toISOString();
    
    // Capture stack trace
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, MCPError);
    }
  }

  /**
   * Convert error to a plain object for logging or serialization
   * @returns {Object} Plain object representation of the error
   */
  toJSON() {
    return {
      name: this.name,
      type: this.type,
      message: this.message,
      severity: this.severity,
      retryable: this.retryable,
      code: this.code,
      details: this.details,
      suggestion: this.suggestion,
      timestamp: this.timestamp
    };
  }

  /**
   * Format the error for display
   * @returns {string} Formatted error message
   */
  format() {
    let output = `[${this.severity.toUpperCase()}] ${this.message} (${this.type})`;
    
    if (this.suggestion) {
      output += `\nSuggestion: ${this.suggestion}`;
    }
    
    if (Object.keys(this.details).length > 0) {
      output += `\nDetails: ${JSON.stringify(this.details, null, 2)}`;
    }
    
    return output;
  }
}

/**
 * Error factory functions for common error types
 */
export const createConnectionError = (message, details = {}, retryable = true) => {
  return new MCPError(message, ErrorType.CONNECTION_FAILED, {
    severity: ErrorSeverity.ERROR,
    retryable,
    code: 503,
    details,
    suggestion: 'Check server status and network connectivity'
  });
};

export const createTimeoutError = (message, details = {}) => {
  return new MCPError(message || 'Operation timed out', ErrorType.TIMEOUT, {
    severity: ErrorSeverity.ERROR,
    retryable: true,
    code: 408,
    details,
    suggestion: 'Try again with a longer timeout or when server load is lower'
  });
};

export const createToolNotFoundError = (toolName, details = {}) => {
  return new MCPError(`Tool '${toolName}' not found`, ErrorType.TOOL_NOT_FOUND, {
    severity: ErrorSeverity.ERROR,
    retryable: false,
    code: 404,
    details,
    suggestion: 'Check the tool name and ensure the tool is registered on the server'
  });
};

export const createParameterError = (message, details = {}) => {
  return new MCPError(message, ErrorType.INVALID_PARAMETERS, {
    severity: ErrorSeverity.ERROR,
    retryable: false,
    code: 400,
    details,
    suggestion: 'Check the parameter values and try again'
  });
};

export const createParsingError = (message, details = {}) => {
  return new MCPError(message || 'Failed to parse response', ErrorType.PARSING_ERROR, {
    severity: ErrorSeverity.ERROR,
    retryable: false,
    code: 500,
    details,
    suggestion: 'Check that the response format matches the expected schema'
  });
};

/**
 * Retry mechanism for retryable errors
 * @param {Function} operation - Async function to retry
 * @param {Object} options - Retry options
 * @param {number} [options.maxRetries=3] - Maximum number of retry attempts
 * @param {number} [options.initialDelay=1000] - Initial delay in milliseconds
 * @param {number} [options.maxDelay=30000] - Maximum delay in milliseconds
 * @param {Function} [options.shouldRetry] - Function to determine if retry should be attempted
 * @returns {Promise<any>} Result of the operation
 */
export const retry = async (operation, options = {}) => {
  const maxRetries = options.maxRetries || 3;
  const initialDelay = options.initialDelay || 1000;
  const maxDelay = options.maxDelay || 30000;
  const shouldRetry = options.shouldRetry || ((error) => error.retryable === true);
  
  let lastError;
  let attempt = 0;
  
  while (attempt <= maxRetries) {
    try {
      return await operation();
    } catch (error) {
      lastError = error;
      attempt += 1;
      
      // Check if we should retry
      if (attempt > maxRetries || !shouldRetry(error)) {
        throw error;
      }
      
      // Calculate backoff delay with exponential backoff and jitter
      const delay = Math.min(
        initialDelay * Math.pow(2, attempt - 1) * (0.5 + Math.random()),
        maxDelay
      );
      
      // Log retry attempt
      console.warn(`Retry attempt ${attempt}/${maxRetries} after ${Math.round(delay)}ms: ${error.message}`);
      
      // Wait before retrying
      await new Promise(resolve => setTimeout(resolve, delay));
    }
  }
  
  // This should never happen, but just in case
  throw lastError;
};

/**
 * Handle errors consistently
 * @param {Error} error - Error to handle
 * @param {Object} options - Handling options
 * @param {boolean} [options.exitOnFatal=true] - Whether to exit process on fatal errors
 * @param {Function} [options.logger=console.error] - Logging function
 * @returns {Object} Standardized error object
 */
export const handleError = (error, options = {}) => {
  const exitOnFatal = options.exitOnFatal !== false;
  const logger = options.logger || console.error;
  
  // Convert to MCPError if it's not already
  const mcpError = error instanceof MCPError 
    ? error 
    : new MCPError(error.message || 'Unknown error occurred', ErrorType.UNKNOWN, {
        severity: ErrorSeverity.ERROR,
        retryable: false,
        code: 500,
        details: { originalError: error.toString() }
      });
  
  // Log the error
  logger(mcpError.format());
  
  // Exit on fatal errors if configured to do so
  if (exitOnFatal && mcpError.severity === ErrorSeverity.FATAL) {
    process.exit(1);
  }
  
  return mcpError.toJSON();
}; 

File: tools/lib/error_handler.py
#!/usr/bin/env python3

"""Error Handler

This module provides standardized error handling for all scripts in the project.

Maturity: beta

Why:
- Consistent error handling across all scripts
- Better error reporting and debugging
- Proper handling of critical vs non-critical errors
- Improved user experience with clear error messages
"""

import sys
import logging
from enum import Enum
from typing import Optional, Dict, Any
from dataclasses import dataclass
from datetime import datetime

class ErrorSeverity(Enum):
    """Error severity levels."""
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"
    CRITICAL = "critical"
    FATAL = "fatal"

class ErrorType(Enum):
    """Error types for categorization."""
    UNKNOWN = "unknown"
    FILE_NOT_FOUND = "file_not_found"
    INVALID_FORMAT = "invalid_format"
    VALIDATION_ERROR = "validation_error"
    PROCESSING_ERROR = "processing_error"
    CONNECTION_ERROR = "connection_error"
    TIMEOUT_ERROR = "timeout_error"
    PERMISSION_ERROR = "permission_error"
    CONFIGURATION_ERROR = "configuration_error"
    DEPENDENCY_ERROR = "dependency_error"

@dataclass
class ErrorDetails:
    """Details about an error."""
    message: str
    type: ErrorType
    severity: ErrorSeverity
    timestamp: datetime
    file: Optional[str] = None
    line: Optional[int] = None
    context: Optional[Dict[str, Any]] = None
    suggestion: Optional[str] = None
    retryable: bool = False

class ErrorHandler:
    """Handles errors consistently across the project."""
    
    def __init__(self, verbose: bool = False, exit_on_fatal: bool = True):
        self.verbose = verbose
        self.exit_on_fatal = exit_on_fatal
        self.logger = logging.getLogger("error_handler")
        
        # Configure logging
        if verbose:
            self.logger.setLevel(logging.DEBUG)
            handler = logging.StreamHandler(sys.stderr)
            handler.setFormatter(logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s'))
            self.logger.addHandler(handler)
    
    def handle_error(self, error: Exception, 
                    error_type: ErrorType = ErrorType.UNKNOWN,
                    severity: ErrorSeverity = ErrorSeverity.ERROR,
                    file: Optional[str] = None,
                    line: Optional[int] = None,
                    context: Optional[Dict[str, Any]] = None,
                    suggestion: Optional[str] = None,
                    retryable: bool = False) -> ErrorDetails:
        """Handle an error and return error details."""
        # Create error details
        details = ErrorDetails(
            message=str(error),
            type=error_type,
            severity=severity,
            timestamp=datetime.now(),
            file=file,
            line=line,
            context=context,
            suggestion=suggestion,
            retryable=retryable
        )
        
        # Log the error
        self._log_error(details)
        
        # Handle based on severity
        if severity == ErrorSeverity.FATAL and self.exit_on_fatal:
            self.logger.error("Fatal error occurred. Exiting...")
            sys.exit(1)
        
        return details
    
    def _log_error(self, details: ErrorDetails) -> None:
        """Log an error with appropriate level and details."""
        log_message = f"{details.type.value}: {details.message}"
        
        if details.file:
            log_message += f" in {details.file}"
            if details.line:
                log_message += f" at line {details.line}"
        
        if details.context:
            log_message += f"\nContext: {details.context}"
        
        if details.suggestion:
            log_message += f"\nSuggestion: {details.suggestion}"
        
        if details.retryable:
            log_message += "\nThis error is retryable."
        
        # Log based on severity
        if details.severity == ErrorSeverity.INFO:
            self.logger.info(log_message)
        elif details.severity == ErrorSeverity.WARNING:
            self.logger.warning(log_message)
        elif details.severity == ErrorSeverity.ERROR:
            self.logger.error(log_message)
        elif details.severity == ErrorSeverity.CRITICAL:
            self.logger.critical(log_message)
        elif details.severity == ErrorSeverity.FATAL:
            self.logger.critical(log_message)
    
    def create_file_not_found_error(self, file_path: str, suggestion: Optional[str] = None) -> ErrorDetails:
        """Create a file not found error."""
        return self.handle_error(
            FileNotFoundError(f"File not found: {file_path}"),
            error_type=ErrorType.FILE_NOT_FOUND,
            severity=ErrorSeverity.ERROR,
            context={"file_path": file_path},
            suggestion=suggestion or f"Check if the file exists at: {file_path}"
        )
    
    def create_invalid_format_error(self, message: str, context: Optional[Dict[str, Any]] = None) -> ErrorDetails:
        """Create an invalid format error."""
        return self.handle_error(
            ValueError(message),
            error_type=ErrorType.INVALID_FORMAT,
            severity=ErrorSeverity.ERROR,
            context=context,
            suggestion="Check the format of the input data"
        )
    
    def create_validation_error(self, message: str, context: Optional[Dict[str, Any]] = None) -> ErrorDetails:
        """Create a validation error."""
        return self.handle_error(
            ValueError(message),
            error_type=ErrorType.VALIDATION_ERROR,
            severity=ErrorSeverity.ERROR,
            context=context,
            suggestion="Verify the input data meets all requirements"
        )
    
    def create_processing_error(self, message: str, context: Optional[Dict[str, Any]] = None) -> ErrorDetails:
        """Create a processing error."""
        return self.handle_error(
            Exception(message),
            error_type=ErrorType.PROCESSING_ERROR,
            severity=ErrorSeverity.ERROR,
            context=context,
            suggestion="Check the processing logic and input data"
        )
    
    def create_connection_error(self, message: str, context: Optional[Dict[str, Any]] = None) -> ErrorDetails:
        """Create a connection error."""
        return self.handle_error(
            ConnectionError(message),
            error_type=ErrorType.CONNECTION_ERROR,
            severity=ErrorSeverity.ERROR,
            context=context,
            suggestion="Check network connectivity and server status",
            retryable=True
        )
    
    def create_timeout_error(self, message: str, context: Optional[Dict[str, Any]] = None) -> ErrorDetails:
        """Create a timeout error."""
        return self.handle_error(
            TimeoutError(message),
            error_type=ErrorType.TIMEOUT_ERROR,
            severity=ErrorSeverity.ERROR,
            context=context,
            suggestion="Try again with a longer timeout or when server load is lower",
            retryable=True
        )
    
    def create_permission_error(self, message: str, context: Optional[Dict[str, Any]] = None) -> ErrorDetails:
        """Create a permission error."""
        return self.handle_error(
            PermissionError(message),
            error_type=ErrorType.PERMISSION_ERROR,
            severity=ErrorSeverity.ERROR,
            context=context,
            suggestion="Check file permissions and user access rights"
        )
    
    def create_configuration_error(self, message: str, context: Optional[Dict[str, Any]] = None) -> ErrorDetails:
        """Create a configuration error."""
        return self.handle_error(
            ValueError(message),
            error_type=ErrorType.CONFIGURATION_ERROR,
            severity=ErrorSeverity.ERROR,
            context=context,
            suggestion="Verify configuration settings and environment variables"
        )
    
    def create_dependency_error(self, message: str, context: Optional[Dict[str, Any]] = None) -> ErrorDetails:
        """Create a dependency error."""
        return self.handle_error(
            ImportError(message),
            error_type=ErrorType.DEPENDENCY_ERROR,
            severity=ErrorSeverity.ERROR,
            context=context,
            suggestion="Check if all required dependencies are installed"
        ) 